crates/rpc/tests/rpc.rs

```diff
@@ -0,0 +1,477 @@
+    block: BlockWithTxs,
+    block_id: BlockId,
+    extracted_value: T,
+}
+
+/// Creates a `TestContext` with the latest block.
+async fn latest_block_context() -> TestContext<()> {
+    context(|_block| Some(())).await
+}
+
+/// Creates a `TestContext` with the latest block with at least N transactions.
+async fn n_txs_context(min_tx_number: usize) -> TestContext<()> {
+    context(|block| {
+        if block.transactions.len() >= min_tx_number {
+            Some(())
+        } else {
+            None
+        }
+    })
+    .await
+}
+
+/// Instantiate a client and look for a block matching the `extractor`, going from latest
+///  to older.
+///
+/// # Parameters
+///
+/// * `extractor`: a closure returning `Some(T)` in the case of a match. The value returned
+///  by the extractor will end up in the `TestContext`'s `extracted_value` field.
+async fn context<F: Fn(&BlockWithTxs) -> Option<T>, T>(
+    extractor: F,
+) -> TestContext<T> {
+    let rpc_url: Url =
+        "http://localhost:3030".parse().expect("Invalid RPC URL");
+    let mut client = TestClient {
+        inner: JsonRpcClient::new(HttpTransport::new(rpc_url)),
+        block_cache: SizedCache::with_size(1000),
+    };
+
+    let latest = client
+        .block_number()
+        .await
+        .expect("Failed to retrieve the latest block number");
+
+    let block = match client
+        .get_cached_block_with_txs(latest)
+        .await
+        .expect("Failed to retrieve the latest block")
+    {
+        MaybePendingBlockWithTxs::Block(block) => block,
+        MaybePendingBlockWithTxs::PendingBlock(_) => {
+            panic!("The latest block shouldn't be a pending block")
+        }
+    };
+
+    if let Some(extracted_value) = extractor(&block) {
+        return TestContext {
+            client,
+            block_id: BlockId::Number(block.block_number),
+            block,
+            extracted_value,
+        };
+    }
+
+    // The latest block doesn't match the criterion.
+    // Go to the lower blocks one by one until a match is found.
+
+    let mut limit = 1_000;
+    let mut block_number = block.block_number;
+
+    while limit > 0 {
+        limit -= 1;
+        block_number -= 1;
+
+        if block_number == 0 {
+            panic!("Reached the genesis block, still no suitable block found");
+        }
+
+        let block = match client
+            .get_cached_block_with_txs(block_number)
+            .await
+            .expect("Block retrieval failed")
+        {
+            MaybePendingBlockWithTxs::Block(block) => block,
+            MaybePendingBlockWithTxs::PendingBlock(_) => {
+                continue;
+            }
+        };
+
+        if let Some(extracted_value) = extractor(&block) {
+            return TestContext {
+                client,
+                block_id: BlockId::Number(block.block_number),
+                block,
+                extracted_value,
+            };
+        }
+    }
+
+    panic!("No suitable block found")
```

Comment:
Explicit `panic!` always freaks me out a bit :)

Another alternative would be returning wrapped result here (`Result` or `Option`), that can be simply unwrapped in the test code (thus pointing to a specific line & test where the problem occurred, not to this line).

Reply:
I get the reluctance to use panics, but I see more boilerplate for not additional value here.